File: 3_Java(eclipse)/1_workspace_ServletAndJSP/1_書籍検索アプリ/src/main/java/todo/model/BookDataSearcher.java
package todo.model;

import java.util.ArrayList;
import java.util.List;

import javax.servlet.http.HttpServletRequest;

import todo.model.BookDataManager.FunctionPointer;
import todo.model.beans.BookDataBean;

public class BookDataSearcher {
	
	private BookDataManager bookDataManager;
	private SearchInformationDataManager searchInformationDataManager;
	
	//----------------
	//コンストラクタ
	public BookDataSearcher(BookDataManager bookDataManager,
		SearchInformationDataManager searchInformationDataManager) {
		
		this.bookDataManager = bookDataManager;
		this.searchInformationDataManager = searchInformationDataManager;
	}
	
	//-----------
	//検索
	public void searchBookData(String keyWord,HttpServletRequest request){
		
		if(!"".equals(keyWord)) {
			
			String pushedRadio = searchInformationDataManager.getPushedRadio();
//			switch(pushedRadio) {
//			
//				case "keyWordSearch":
//					//キーワードで検索
//					keyWordSearch(pushedRadio,keyWord);
//				break;
//				case "titleSearch":
//					//本のタイトルで検索
//					bookTitleNameSearch(pushedRadio,keyWord);
//				break;
//			}
			keyWordSearch(pushedRadio,keyWord);
			searchInformationDataManager.pushSearchButton(true,request);
		}else {
			
			searchInformationDataManager.pushSearchButton(false,request);
		}
	}
	
	//------------
	//検索
	public ArrayList<BookDataBean> keyWordSearch(String pushedRadio,String keyWord) {
		
		String[] chkBoxDataArray = searchInformationDataManager.getPushedChkBox();
		BookDataBean[] beanData = bookDataManager.mergedBookDataBeanGet(chkBoxDataArray); 
		var searchResultList = 
			"keyWordSearch".equals(pushedRadio)? 
				//キーワードで検索
				(ArrayList<BookDataBean>)keyWordSearchDetail(beanData,keyWord)
				//本のタイトルで検索
				:(ArrayList<BookDataBean>)bookTitleNameSearchDetail(beanData,keyWord);
		searchInformationDataManager.setSearchResultList(searchResultList);
		return searchResultList;
	}
	
//	//------------
//	//本のタイトルで検索
//	public ArrayList<BookDataBean> bookTitleNameSearch(String titleName) {
//		
//		var searchResultList = 
//			(ArrayList<BookDataBean>)bookTitleNameSearchDetail(bookDataManager.getBookDataBeans(),titleName);
//		searchInformationDataManager.setSearchResultList(searchResultList);
//		return searchResultList;
//	}
//	
	private List<BookDataBean> keyWordSearchDetail(BookDataBean[] bookDataBeans,String keyWord) {
		
		List<BookDataBean> searchResultList = new ArrayList<>();
		for(FunctionPointer search:bookDataManager.SEARCH_TARGET_ARRAY) {
			
			for(BookDataBean bookData : bookDataBeans) {
				
				if(bookDataManager.stringRegexCheck(search.apply(bookData),keyWord)) {
					
					searchResultList.add(bookData);
				}
			}	
		}
		return searchResultList;
	}
	
	private List<BookDataBean> bookTitleNameSearchDetail(BookDataBean[] bookDataBeans,String bookTitleName) {
		
		List<BookDataBean> searchResultList = new ArrayList<>();
		for(BookDataBean bookData : bookDataBeans) {
			
			if(bookDataManager.stringRegexCheck(bookDataManager.getBookName(bookData),bookTitleName)) {
				
				searchResultList.add(bookData);
			}
		}	
		return searchResultList;
	}
}

File: 1_C#(VisualStudio)/1_workspace_ASP.NETCoreMVC/1_書籍検索アプリ_移植元Java/Models/BookDataSearcher.cs
using 書籍検索アプリ_移植元Java.Models.beans;

namespace 書籍検索アプリ_移植元Java.Models {
    public class BookDataSearcher {

        private readonly BookDataManager bookDataManager;
        // 検索対象プロパティを関数として配列化
        private readonly Func<BookDataBean, string>[] searchTargets =
        [
			//検索は上のものから順に探す
			//本のタイトルで検索
			book => book.BookName!,
			//著者で検索
			book => book.Author!,
			//架空ISBNコードで検索
			book => book.FictitiousISBN!,
        ];

        //----------------
        //コンストラクタ
        public BookDataSearcher(BookDataManager bookDataManager) {

            this.bookDataManager = bookDataManager;
        }

        //-----------
        //検索
        public List<BookDataBean>? SearchBookData(string? keyWord, string? pushedRadio, string[]? chkBoxDataArray) {

            if (keyWord == null || pushedRadio == null || chkBoxDataArray == null || keyWord == "") {

                return null;
            }
            return BookDataSearch(pushedRadio, keyWord, chkBoxDataArray);
        }

        //------------
        //検索
        public List<BookDataBean> BookDataSearch(string pushedRadio, string keyWord, string[] chkBoxDataArray) {

            BookDataBean[] beanData = bookDataManager.MergedBookDataBeanGet(chkBoxDataArray);
            var searchResultList =
                pushedRadio == "keyWordSearch" ?
                    //キーワードで検索
                    (List<BookDataBean>)KeyWordSearch(beanData, keyWord)
                    //本のタイトルで検索
                    : (List<BookDataBean>)BookTitleNameSearch(beanData, keyWord);
            return searchResultList;
        }

        private List<BookDataBean> KeyWordSearch(BookDataBean[] bookDataBeans, string keyWord) {

            return BookDataSearchDetail(bookDataBeans, keyWord);
        }

        private List<BookDataBean> BookTitleNameSearch(BookDataBean[] bookDataBeans, string bookTitleName) {

            return BookDataSearchDetail(bookDataBeans, bookTitleName, 1);
        }

        private List<BookDataBean> BookDataSearchDetail(BookDataBean[] bookDataBeans, string keyWord, int end = 2) {

            return searchTargets
                .Take(end) // end より後ろの検索対象をスキップ
                .SelectMany(selector =>
                    bookDataBeans.Where(bookData =>
                        bookDataManager.StringRegexCheck(selector(bookData), keyWord)))
                .ToList();
        }
    }
}
